website/docs/opengl/c1/2-hello-quad.md

```diff
@@ -0,0 +1,524 @@
+
+### Filling our buffer with data
+Let's fill our buffer with some data! Before we do that though, you need to be aware of `unsafe` in C#.
+
+#### Unsafe C#
+Silk.NET heavily uses `unsafe` code. Don't worry, this won't make your computer explode, however it does exit out of the "safe" managed environment of C#, and enters a realm where undefined behaviour, segmentation faults, and strange results are more likely to occur if you are not careful.
+
+<?# Info "If you wish to use `Span` instead, and remain in `safe` mode, Silk.NET does support these too. However, I will be using `unsafe` in this tutorial instead, as this is both what I personally use, as well as what the samples use." /?>
+
+Unsafe mode is not enabled by default, so we need to enable it. To enable it, add `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>` to the `PropertyGroup` in your project's `.csproj` file. You also **MUST** add `unsafe` to your `OnLoad` and `OnRender` methods, like so:
+
+```cs
+public static unsafe void OnLoad() { ... }
+```
+
+If you forget to do either of these, you'll get compile errors. While our `OnRender` method does not contain any unsafe code yet, it will when we draw to the screen later, so it's best just to do it now so you don't forget to do it later.
+
+Now let's fill the buffer!
+
+Add the following after you create the buffer:
+
+```cs
+fixed (float* buf = vertices)
+    _gl.BufferData(BufferTargetARB.ArrayBuffer, (nuint) (vertices.Length * sizeof(float)), buf, BufferUsageARB.StaticDraw);
+```
+
+Let's go over what's going on here. First, we `fix` the vertices data. This prevents the garbage collector from moving it around, so we can take a pointer to the data. We tell it the target we want, the `ArrayBuffer` in this case, give it the data length, the buffer pointer, and we choose `StaticDraw` for our usage.
+
+Now, run the program again, and if you get the same blue window, you've set the buffer data successfully!
+
+Now, let's create the **element buffer object**.
+
+## Element Buffer Objects (EBOs)
+Unlike a vertex buffer, an element buffer is not *strictly* required to display something on screen, however not using it will require making your vertex buffer a lot larger, with a lot more duplicate data.
+
+For those familliar with Direct3D terms, EBOs are more commonly referred to as **index buffers**, which is what this tutorial will refer to them from now on.
+
+Take a look at the `vertices` array we defined earlier. You may have noticed that it only has four different points in it. Great! That's how you make a quad. But remember - that's not how this works. As mentioned earlier, a quad is made up of two triangles. But wait - that's six points... But we've only defined four. What's going on?
+
+Well, that's where index buffers come in! An index buffer simply tells the GPU which points in the vertex buffer to use for each triangle.
+
+Add the following to your `OnLoad` method:
+
+```cs
+uint[] indices =
+{
+    0u, 1u, 3u,
+    1u, 2u, 3u
+};
+```
```

Comment:
Any particular reason for using `uint` for indices? I usually use `ushort` or if that's not enough `uint`, given that those are _always_ 16-bit and 32-bit respectively

Reply:
I think its best to suggest `ushort` by default, as its half the data

Reply:
Okay yes, I'm gonna ask you to change it because afterwards you have to specify the element type, and you do so as `UnsignedInt`, therefore you need to make sure that your indices are actually 32-bit.

You can change it to `uint` if you wish, I'd still recommend `ushort`. You'd also have to update part of the "Drawing to the screen" section.

Reply:
I personally use uint for my indices which is why I used it here, but it makes little difference in practice. The main reason I use uint is to support more than 65k vertices, which some of the models I import have. Personally for the purposes of the tutorial I don't think it makes much difference but I'm happy to change it.

Reply:
Ah sorry my bad; I got confused and somehow thought you were using native-sized integers (nuint). I'm a big dum dum. I'll mark this as resolved